GZY-437: Restrict task visibility to employee’s own projects in ERP

vestefano opened the pull request:
- This change enforces a consistent project-visibility rule across the system, ensuring that both Managers and Employees follow the same assigned-projects restriction.

- The previous behavior allowed Managers to see all projects by default, which conflicted with the intended access-control model and caused inconsistencies in how visibility rules were applied across different endpoints.

- To maintain coherence with the platform's permission model and avoid privilege gaps between roles, the assigned-projects rule was formally applied to Employees as well.

- Backend services that return projects or project-related tasks were updated to ensure permission checks are applied uniformly, preventing accidental exposure of unassigned projects.

- A migration was added to guarantee that roles have the correct permission associations, avoiding environments where outdated role-permission links could break the new visibility logic.

- Minor frontend adjustments were made to fix the employee filter in the project-management dashboard, since incorrect parameter mapping caused the API to receive invalid filter values and could misapply visibility rules.

**Card:** [Restrict Task Visibility to Employee’s Own Projects in ERP](https://trello.com/c/EhgJ3STb/437-restrict-task-visibility-to-employees-own-projects-in-erp?filter=member%3Avictoralfonsoestefanoroman)

vestefano committed:
feat: enforce consistent project visibility rules across roles

- This change enforces a consistent project-visibility rule across the system, ensuring that both
Managers and Employees follow the same assigned-projects restriction.

- The previous behavior allowed Managers to see all projects by default, which conflicted with the
intended access-control model and caused inconsistencies in how visibility rules were applied
across different endpoints.

- To maintain coherence with the platform's permission model and avoid privilege gaps between
roles, the assigned-projects rule was formally applied to Employees as well.

- Backend services that return projects or project-related tasks were updated to ensure permission
checks are applied uniformly, preventing accidental exposure of unassigned projects.

- A migration was added to guarantee that roles have the correct permission associations, avoiding
environments where outdated role-permission links could break the new visibility logic.

- Minor frontend adjustments were made to fix the employee filter in the project-management
dashboard, since incorrect parameter mapping caused the API to receive invalid filter values and
could misapply visibility rules.

YanaYestr commented on packages/core/src/lib/role-permission/default-role-permissions.ts:
```diff
@@ -500,6 +500,7 @@ export const DEFAULT_ROLE_PERMISSIONS = [
 			PermissionsEnum.PROJECT_MODULE_UPDATE,
 			PermissionsEnum.PROJECT_MODULE_DELETE,
+			PermissionsEnum.VIEW_ASSIGNED_PROJECTS_ONLY,
 			/** Dashboard */
```
If you use that permission for employees, then you also need to update the translation so it no longer says “for managers only.” Can you please change that?

YanaYestr commented on apps/gauzy/src/app/pages/dashboard/project-management/project-management-details/project-management-details.component.ts:
```diff
@@ -109,7 +109,7 @@ export class ProjectManagementDetailsComponent extends PaginationFilterBaseCompo
 				organizationId,
 				tenantId,
-				...(this.selectedEmployeeId ? { employeeId: this.selectedEmployeeId } : {}),
+				...(this.selectedEmployeeId ? { members: { id: this.selectedEmployeeId } } : {}),
```
For managers, I can see data from other projects as well, not only from the one I’m assigned to.
<img width="1057" height="754" alt="Zrzut ekranu 2025-11-20 o 12 35 35" src="https://github.com/user-attachments/assets/e6c2a446-3c36-48eb-856a-0d7bb57cbe01" />

YanaYestr commented on packages/core/src/lib/tasks/task.service.ts:
```diff
@@ -492,84 +529,66 @@ export class TaskService extends TenantAwareCrudService<Task> {
-					}
-				} else if (isManager) {
-					const selectedEmployeeId = members?.['id'];
-					const managerId = RequestContext.currentUser().employeeId;
```
<img width="1117" height="580" alt="Zrzut ekranu 2025-11-20 o 12 26 24" src="https://github.com/user-attachments/assets/14754b46-1f02-4006-be41-ffd5cf3b4170" />
And in Tasks, as a manager, I also see data from projects I'm not assigned to, not only the ones I'm added to.

YanaYestr replied:
Also, for the employee, the tasks from their project are not visible.

https://github.com/user-attachments/assets/af988269-ffb7-42bb-b558-825c8e496da7

YanaYestr reviewed (changes requested):
And as we discussed, the remaining screens - such as Timesheets and Reports -will be done under this task, but in a separate branch. It’s based on what is described in the ticket 
<img width="629" height="106" alt="Zrzut ekranu 2025-11-20 o 13 05 49" src="https://github.com/user-attachments/assets/55a03047-ca0e-4255-b7c6-0195a17de092" />

YanaYestr commented on packages/core/src/lib/tasks/task.service.ts:
```diff
@@ -868,8 +946,49 @@ export class TaskService extends TenantAwareCrudService<Task> {
-		const filters = options?.filters;
-		const where = options?.where;
+		const { where, filters } = options;
+
```
Please fix error with admin permission error 

https://github.com/user-attachments/assets/7324a154-1814-4103-96ca-88b1d2964d66

YanaYestr replied:
So not all available tasks are shown here - there are many more of them in the database.

vestefano replied:
When an admin user is treated as an employee and accesses the task dashboard, and tries to filter by project, the request sends the organizationSprintId parameter with the value 'null' as a string. This causes that value to be used as a filter in the query, which is why the error shown in the video occurs.

To fix this issue, a condition was added so that if this happens, the value will be converted to null, but using the approach required by TypeORM.

YanaYestr commented on apps/gauzy/src/app/pages/dashboard/project-management/project-management-details/project-management-details.component.ts:
```diff
@@ -109,7 +109,7 @@ export class ProjectManagementDetailsComponent extends PaginationFilterBaseCompo
 				organizationId,
 				tenantId,
-				...(this.selectedEmployeeId ? { employeeId: this.selectedEmployeeId } : {}),
+				...(this.selectedEmployeeId ? { members: { id: this.selectedEmployeeId } } : {}),
```
Please move all changes that are not related to this task https://trello.com/c/EhgJ3STb/437-restrict-task-visibility-to-employees-own-projects-in-erp to a separate branch. Also, please review the task and the comments.

vestefano replied:
This change is related to this task; without it, filtering by employee would not be possible, for example.

YanaYestr commented on packages/core/src/lib/organization-project/organization-project.service.ts:
```diff
@@ -445,6 +446,37 @@ export class OrganizationProjectService extends TenantAwareCrudService<Organizat
 			query.andWhere(`project_team.id = :organizationTeamId`, { organizationTeamId });
 		}
 
+		const userEmployeeId = RequestContext.currentUser().employeeId;
```
Also, previously I was able to see all projects as an admin, regardless of the filter, but now I can see only the ones that this employee belongs to. So if I want to add them to another project, I won’t be able to. This needs clarification in the requirements.


https://github.com/user-attachments/assets/83e837d3-1202-46cd-b050-54399becd027

YanaYestr replied:
Please add update in comments to that task, that all persons will know that it's desired behavior

vestefano replied:
<img width="519" height="629" alt="image" src="https://github.com/user-attachments/assets/95e52f84-9b9a-4345-8528-afb457e1adb9" />

YanaYestr commented on packages/core/src/lib/tasks/task.service.ts:
```diff
@@ -754,46 +847,54 @@ export class TaskService extends TenantAwareCrudService<Task> {
 			const { teams = [] } = where;
 			const { members } = where;
 
-			const hasPermission =
```
Please move all changes that are not related to this task https://trello.com/c/EhgJ3STb/437-restrict-task-visibility-to-employees-own-projects-in-erp to a separate branch. Also, please review the task and the comments.

vestefano replied:
The changes related to the definition of Manager permissions are directly related to this task. Those changes should be kept in this PR.

YanaYestr commented on packages/core/src/lib/tasks/task.service.ts:
```diff
@@ -1122,9 +1217,45 @@ export class TaskService extends TenantAwareCrudService<Task> {
 			const { modules = [], members } = where;
 
+			if (RequestContext.hasPermission(PermissionsEnum.VIEW_ASSIGNED_PROJECTS_ONLY)) {
+				const userEmployeeId = RequestContext.currentUser().employeeId;
```
Please move all changes that are not related to this task https://trello.com/c/EhgJ3STb/437-restrict-task-visibility-to-employees-own-projects-in-erp to a separate branch. Also, please review the task and the comments.

YanaYestr commented on packages/core/src/lib/tasks/task.service.ts:
```diff
@@ -675,30 +731,67 @@ export class TaskService extends TenantAwareCrudService<Task> {
 		options: PaginationParams<Task> & IAdvancedTaskFiltering
 	) {
 		try {
+			const { filters, where } = options;
```
For the employee, I don’t see any added projects or tasks anywhere.

https://github.com/user-attachments/assets/84437863-ffa6-46e7-8b90-f79064231b78

YanaYestr reviewed (changes requested):
Please, could you check the functionality for all roles once again, and move all code that is not related to these three tasks to another branch: https://trello.com/c/EhgJ3STb/437-restrict-task-visibility-to-employees-own-projects-in-erp
, https://trello.com/c/KtABlgdK/470-restrict-task-visibility-to-employees-own-projects-in-teams-task-dashboard
, https://trello.com/c/rkN2zxqo/469-restrict-task-visibility-to-employees-own-projects-in-my-tasks-dashboard.

YanaYestr commented on packages/core/src/lib/core/crud/tenant-aware-crud.service.ts:
```diff
@@ -28,16 +28,17 @@ export abstract class TenantAwareCrudService<T extends TenantBaseEntity>
 	 */
 	private findConditionsWithEmployeeByUser(): FindOptionsWhere<T> {
-		const employeeId = RequestContext.currentEmployeeId();
+		const employeeId = RequestContext.currentUser().employeeId;
```
There are very large changes here that affect the entire app. Please check whether it’s possible to apply the changes only in the places that are related to tasks.

YanaYestr commented on packages/core/src/lib/core/crud/tenant-aware-crud.service.ts:
```diff
@@ -28,16 +28,17 @@ export abstract class TenantAwareCrudService<T extends TenantBaseEntity>
-					  this.typeOrmRepository.metadata?.hasColumnWithPropertyPath('employeeId')
+					? (!RequestContext.hasPermission(PermissionsEnum.CHANGE_SELECTED_EMPLOYEE)
+					  || RequestContext.hasPermission(PermissionsEnum.VIEW_ASSIGNED_PROJECTS_ONLY))
+					  && this.typeOrmRepository.metadata?.hasColumnWithPropertyPath('employeeId')
```
the same. There are very large changes here that affect the entire app. Please check whether it’s possible to apply the changes only in the places that are related to tasks.

vestefano committed:
revert: changes in TenantAwareCrudService

This change was reverted because it was outside the scope of the current task.

vestefano committed:
chore: revert task service changes

- Reverts changes in task service because it are out of scope for this PR.

vestefano committed:
fix: handle null organizationSprintId in task filtering

- Updated the task service to treat 'null' as a valid case for setting
  organizationSprintId to null in the filter options.
